File: android/src/main/kotlin/com/ko2ic/fluttergoogleadmanager/NativeAd.kt
package com.ko2ic.fluttergoogleadmanager

import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.doubleclick.PublisherAdRequest
import com.google.android.gms.ads.formats.NativeCustomTemplateAd
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

class NativeAd(private val registrar: PluginRegistry.Registrar, private val channel: MethodChannel) : MethodChannel.MethodCallHandler {
    private lateinit var builder: AdLoader.Builder
    private lateinit var ad: NativeCustomTemplateAd

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "load" -> load(call, result)
            "dispose" -> dispose(result)
            "performClick" -> performClick(call, result)
            "getParameter" -> getParameter(call, result)
            else -> result.notImplemented()
        }
    }

    private fun load(call: MethodCall, result: MethodChannel.Result) {
        val isDevelop = call.argument<Boolean>("isDevelop") ?: false
        if (isDevelop) {
            builder = AdLoader.Builder(registrar.context(), "/6499/example/native")
                    .forCustomTemplateAd("10063170",
                            { ad ->
                                this.ad = ad
                                this.ad.recordImpression()
                            }, null)
        } else {
            val adUnit = call.argument<String>("adUnitId")
            val templateId = call.argument<String>("templateId")
            builder = AdLoader.Builder(registrar.context(), adUnit)
                    .forCustomTemplateAd(templateId,
                            { ad ->
                                this.ad = ad
                                this.ad.recordImpression()
                            }, null)
        }
        val adRequestBuilder = PublisherAdRequest.Builder()
        val customTargeting = call.argument("customTargeting") as? Map<String, Any>

        customTargeting?.let {
            it.entries.forEach { (key, value) ->
                when(value){
                    is String -> adRequestBuilder.addCustomTargeting(key, value)
                    is List<*> -> adRequestBuilder.addCustomTargeting(key, value.filterIsInstance<String>())
                    else -> throw IllegalArgumentException("customTargeting: values must be either Strings or Lists of Strings, but got $value")
                }
            }
        }
        val adLoader = builder.withAdListener(NativeAdListener(channel))
                .build()
        adLoader.loadAd(adRequestBuilder.build())

        result.success(null)
    }

    private fun dispose(result: MethodChannel.Result) {
        ad.destroy()
        result.success(null)
    }

    private fun performClick(call: MethodCall, result: MethodChannel.Result) {
        val parameter = call.argument<String>("parameter")
        if (parameter != null) {
            ad.performClick(parameter)
            result.success(null)
        } else {
            result.error("MISSING_PARAMETER", "You must specify the parameter to perform this action", null)
        }
    }

    private fun getParameter(call: MethodCall, result: MethodChannel.Result) {
        val type = call.argument<String>("type")
        val parameter = call.argument<String>("parameter")

        if (type != null) {
            if (parameter != null) {
                when (type) {
                    "image" -> result.success(ad.getImage(parameter).uri.toString())
                    "text" -> result.success(ad.getText(parameter).toString())
                    else -> result.error("INVALID_TYPE", "The type you specified is invalid", null)
                }
            } else {
                result.error("MISSING_PARAMETER", "You must specify the parameter you want", null)
            }
        } else {
            result.error("MISSING_TYPE", "You must send the type of parameter", null)
        }
    }

    class NativeAdListener(private val channel: MethodChannel) : AdListener() {
        override fun onAdLoaded() {
            super.onAdLoaded()
            channel.invokeMethod("onAdLoaded", null)
        }

        override fun onAdFailedToLoad(errorCode: Int) {
            channel.invokeMethod("onAdFailedToLoad", mapOf("errorCode" to errorCode))
        }

        override fun onAdOpened() {
            super.onAdOpened()
            channel.invokeMethod("onAdOpened", null)
        }

        override fun onAdClosed() {
            super.onAdClosed()
            channel.invokeMethod("onAdClosed", null)
        }

        override fun onAdLeftApplication() {
            super.onAdLeftApplication()
            channel.invokeMethod("onAdLeftApplication", null)
        }
    }
}

File: ios/Classes/NativeAd.swift
//
//  NativeAd.swift
//  flutter_google_ad_manager
//
//  Created by Alvaro De la Cruz on 20/11/20.
//

import Foundation
import GoogleMobileAds

class NativeAd: SwiftFlutterGoogleAdManagerPlugin {
    
    private var templateId: String!
    private var adLoader: GADAdLoader!
    
    private var ad: GADNativeCustomTemplateAd?
    private let channel: FlutterMethodChannel!
    
    
    public init(with _: FlutterPluginRegistrar, channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
        channel.setMethodCallHandler({
            (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            self.handle(call, result: result)
        })
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "load":
            load(call, result: result)
        case "getParameter":
            getParameter(call, result: result)
        case "performClick":
            performClick(call, result: result)
        case "dispose":
            dispose(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func load(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let argument = call.arguments as! Dictionary<String, Any>
        let isDevelop = argument["isDevelop"] as? Bool ?? false
        let viewController = UIApplication.shared.delegate!.window!!.rootViewController!
        
        if isDevelop {
            let adUnit = "/6499/example/native"
            self.templateId = "10063170"
            
            adLoader = GADAdLoader(
                adUnitID: adUnit, rootViewController: viewController, adTypes: [ GADAdLoaderAdType.nativeCustomTemplate ], options: [])
        }else{
            let adUnit = argument["adUnitId"] as? String ?? ""
            self.templateId = argument["templateId"] as? String ?? ""
            
            adLoader = GADAdLoader(
                adUnitID: adUnit, rootViewController: viewController, adTypes: [ GADAdLoaderAdType.nativeCustomTemplate ], options: [])
        }
        let customTargeting = argument["customTargeting"] as? [String: Any]
        let request = DFPRequest()
        
        request.customTargeting = customTargeting
        
        adLoader.delegate = self
        adLoader.load(request)
        
        result(nil)
    }
    
    private func performClick(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let nativeAd = ad else {
            result(FlutterError(code: "program_error", message: "Please call load() method first.", details: nil))
            return
        }
        let argument = call.arguments as! Dictionary<String, Any>
        if let parameter = argument["parameter"] as? String {
            nativeAd.performClickOnAsset(withKey: parameter)
        }else{
            result(FlutterError(code: "MISSING_PARAMETER", message: "You must specify the parameter to perform this action", details: nil))
        }
        
    }
    
    private func getParameter(_ call :FlutterMethodCall, result: @escaping FlutterResult){
        guard let nativeAd = ad else {
            result(FlutterError(code: "program_error", message: "Please call load() method first.", details: nil))
            return
        }
        let argument = call.arguments as! Dictionary<String, Any>
        if let type = argument["type"] as? String, let parameter = argument["parameter"] as? String {
            switch(type){
            case "image":
                result(nativeAd.image(forKey: parameter)?.imageURL?.absoluteString)
            case "text":
                result(nativeAd.string(forKey: parameter))
            default:
                result(FlutterError(code: "INVALID_TYPE", message: "The type you specified is invalid", details: nil))
            }
        }else{
            result(FlutterError(code: "MISSING_PARAMETER", message: "You must specify the parameter you want", details: nil))
        }
    }
    
    private func dispose(_: FlutterMethodCall, result: @escaping FlutterResult) {
        result(nil)
    }
}

extension NativeAd: GADNativeCustomTemplateAdLoaderDelegate {
    
    func nativeCustomTemplateIDs(for adLoader: GADAdLoader) -> [String] {
        return [templateId]
    }
    
    func adLoader(_ adLoader: GADAdLoader, didReceive nativeCustomTemplateAd: GADNativeCustomTemplateAd) {
        channel.invokeMethod("onAdLoaded", arguments: nil)
        
        self.ad = nativeCustomTemplateAd
        self.ad?.recordImpression()
    }
    
    func adLoader(_ adLoader: GADAdLoader, didFailToReceiveAdWithError error: GADRequestError) {
        channel.invokeMethod("onAdFailedToLoad", arguments: ["errorCode": error.code])
    }
    
    func nativeAdWillPresentScreen(_ nativeAd: GADNativeAd) {
        channel.invokeMethod("onAdOpened", arguments: nil)
    }
    
    func nativeAdDidDismissScreen(_ nativeAd: GADNativeAd) {
        channel.invokeMethod("onAdClosed", arguments: nil)
    }
    
    func nativeAdWillLeaveApplication(_ nativeAd: GADNativeAd) {
        channel.invokeMethod("onAdLeftApplication", arguments: nil)
    }
}
